### PR TITLE
[ConstraintSystem] Add a property-wrapper constraint to allow for inference of implicit property wrappers

### PR DESCRIPTION
**Pull request by hborla**
Another handful of commits from https://github.com/apple/swift/pull/36344 that are independent from the revision.

* Add a property-wrapper constraint that is used to delay checking if an inferred property-wrapper type is actually a property wrapper.
* Improve diagnostics for invalid use of the `$` syntax with wrapped parameters and arguments.

Resolves: rdar://71902161

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please smoke test

File: lib/Sema/CSBindings.cpp [hunk 1]
```diff
@@ -1251,6 +1251,7 @@ void PotentialBindings::infer(Constraint *constraint) {
   case ConstraintKind::FunctionInput:
   case ConstraintKind::FunctionResult:
   case ConstraintKind::OpaqueUnderlyingType:
+  case ConstraintKind::PropertyWrapper:
```
Comment (xedin) [inline lib/Sema/CSBindings.cpp:1254]: Just like for member references we'd need to make sure that `wrappedValueType` is not bound until `wrapperType` is resolved. You'd have to adjust inference logic and add `PropertyWrapper` constraint to delaying list for the wrapped value e.g. https://github.com/apple/swift/pull/36568/files#diff-f5d7e33c5b6e1d19ce061b6b45945fb3dd74943f11d2e7d61a6886d26405e9f6R1317

**Review by xedin — APPROVED** [review]
LGTM! I have left a comment about inference inline.

File: lib/Sema/CSBindings.cpp (hunk 1 above)
Comment (hborla) [reply lib/Sema/CSBindings.cpp]: Whoops, you're right, thank you! This now behaves the same way as the other member constraints (https://github.com/apple/swift/pull/36568/commits/ee2a8fce16c34df182850074a8b1d5fecb69dab0)

Comment (xedin) [reply lib/Sema/CSBindings.cpp]: Great!

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please smoke test Linux platform

**Pull request merged**